File: src/algorithms/matrices/layer-operations/rotate-layer-by-layer/sources/rotate-layer-by-layer_test.py
import importlib
import copy

rotate_layer_by_layer_mod = importlib.import_module("rotate-layer-by-layer")
rotate_layer_by_layer = rotate_layer_by_layer_mod.rotate_layer_by_layer


def test_rotates_3x3_90_clockwise():
    matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    assert rotate_layer_by_layer(copy.deepcopy(matrix)) == [[7, 4, 1], [8, 5, 2], [9, 6, 3]]


def test_rotates_4x4_90_clockwise():
    matrix = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 16]]
    assert rotate_layer_by_layer(copy.deepcopy(matrix)) == [
        [13, 9, 5, 1],
        [14, 10, 6, 2],
        [15, 11, 7, 3],
        [16, 12, 8, 4],
    ]


def test_handles_1x1_matrix():
    assert rotate_layer_by_layer([[42]]) == [[42]]


def test_rotates_2x2_90_clockwise():
    matrix = [[1, 2], [3, 4]]
    assert rotate_layer_by_layer(copy.deepcopy(matrix)) == [[3, 1], [4, 2]]


def test_rotates_5x5_90_clockwise():
    matrix = [
        [1, 2, 3, 4, 5],
        [6, 7, 8, 9, 10],
        [11, 12, 13, 14, 15],
        [16, 17, 18, 19, 20],
        [21, 22, 23, 24, 25],
    ]
    assert rotate_layer_by_layer(copy.deepcopy(matrix)) == [
        [21, 16, 11, 6, 1],
        [22, 17, 12, 7, 2],
        [23, 18, 13, 8, 3],
        [24, 19, 14, 9, 4],
        [25, 20, 15, 10, 5],
    ]


def test_four_rotations_return_original():
    original = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    matrix = copy.deepcopy(original)
    for _ in range(4):
        matrix = rotate_layer_by_layer(matrix)
    assert matrix == original


def test_handles_negative_and_zero_values():
    matrix = [[-1, 0, 1], [-2, 0, 2], [-3, 0, 3]]
    assert rotate_layer_by_layer(copy.deepcopy(matrix)) == [
        [-3, -2, -1],
        [0, 0, 0],
        [3, 2, 1],
    ]


if __name__ == "__main__":
    test_rotates_3x3_90_clockwise()
    test_rotates_4x4_90_clockwise()
    test_handles_1x1_matrix()
    test_rotates_2x2_90_clockwise()
    test_rotates_5x5_90_clockwise()
    test_four_rotations_return_original()
    test_handles_negative_and_zero_values()
    print("All tests passed!")

File: src/algorithms/matrices/layer-operations/rotate-layer-by-layer/RotateLayerByLayer_test.java
// javac RotateLayerByLayer.java RotateLayerByLayer_test.java && java -ea RotateLayerByLayer_test

import java.util.Arrays;

public class RotateLayerByLayer_test {

    static int[][] deepCopy(int[][] matrix) {
        int[][] copy = new int[matrix.length][];
        for (int rowIdx = 0; rowIdx < matrix.length; rowIdx++) {
            copy[rowIdx] = matrix[rowIdx].clone();
        }
        return copy;
    }

    public static void main(String[] args) {
        testRotates3x3_90Clockwise();
        testRotates4x4_90Clockwise();
        testHandles1x1Matrix();
        testRotates2x2_90Clockwise();
        testRotates5x5_90Clockwise();
        testFourRotationsReturnOriginal();
        testHandlesNegativeAndZeroValues();
        System.out.println("All tests passed!");
    }

    static void testRotates3x3_90Clockwise() {
        int[][] matrix = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
        int[][] result = RotateLayerByLayer.rotateLayerByLayer(deepCopy(matrix));
        assert Arrays.equals(result[0], new int[]{7, 4, 1}) : "Row 0 wrong";
        assert Arrays.equals(result[1], new int[]{8, 5, 2}) : "Row 1 wrong";
        assert Arrays.equals(result[2], new int[]{9, 6, 3}) : "Row 2 wrong";
    }

    static void testRotates4x4_90Clockwise() {
        int[][] matrix = {{1, 2, 3, 4}, {5, 6, 7, 8}, {9, 10, 11, 12}, {13, 14, 15, 16}};
        int[][] result = RotateLayerByLayer.rotateLayerByLayer(deepCopy(matrix));
        assert Arrays.equals(result[0], new int[]{13, 9, 5, 1}) : "Row 0 wrong";
        assert Arrays.equals(result[3], new int[]{16, 12, 8, 4}) : "Row 3 wrong";
    }

    static void testHandles1x1Matrix() {
        int[][] matrix = {{42}};
        int[][] result = RotateLayerByLayer.rotateLayerByLayer(matrix);
        assert result[0][0] == 42;
    }

    static void testRotates2x2_90Clockwise() {
        int[][] matrix = {{1, 2}, {3, 4}};
        int[][] result = RotateLayerByLayer.rotateLayerByLayer(deepCopy(matrix));
        assert Arrays.equals(result[0], new int[]{3, 1}) : "Row 0 wrong";
        assert Arrays.equals(result[1], new int[]{4, 2}) : "Row 1 wrong";
    }

    static void testRotates5x5_90Clockwise() {
        int[][] matrix = {
            {1, 2, 3, 4, 5},
            {6, 7, 8, 9, 10},
            {11, 12, 13, 14, 15},
            {16, 17, 18, 19, 20},
            {21, 22, 23, 24, 25},
        };
        int[][] result = RotateLayerByLayer.rotateLayerByLayer(deepCopy(matrix));
        assert Arrays.equals(result[0], new int[]{21, 16, 11, 6, 1}) : "Row 0 wrong";
        assert Arrays.equals(result[4], new int[]{25, 20, 15, 10, 5}) : "Row 4 wrong";
    }

    static void testFourRotationsReturnOriginal() {
        int[][] original = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
        int[][] matrix = deepCopy(original);
        for (int rotationCount = 0; rotationCount < 4; rotationCount++) {
            matrix = RotateLayerByLayer.rotateLayerByLayer(matrix);
        }
        for (int rowIdx = 0; rowIdx < original.length; rowIdx++) {
            assert Arrays.equals(matrix[rowIdx], original[rowIdx]) : "Row " + rowIdx + " mismatch after 4 rotations";
        }
    }

    static void testHandlesNegativeAndZeroValues() {
        int[][] matrix = {{-1, 0, 1}, {-2, 0, 2}, {-3, 0, 3}};
        int[][] result = RotateLayerByLayer.rotateLayerByLayer(deepCopy(matrix));
        assert Arrays.equals(result[0], new int[]{-3, -2, -1}) : "Row 0 wrong";
        assert Arrays.equals(result[1], new int[]{0, 0, 0}) : "Row 1 wrong";
        assert Arrays.equals(result[2], new int[]{3, 2, 1}) : "Row 2 wrong";
    }
}
